internal/command/gw/gw.go

```diff
@@ -9,42 +9,55 @@ import (
 	"sync"
 
 	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
-	"github.com/spf13/cobra"
+	"github.com/urfave/cli/v2"
+	"github.com/urfave/cli/v2/altsrc"
 	"google.golang.org/grpc"
 
 	mvmv1 "github.com/weaveworks/flintlock/api/services/microvm/v1alpha1"
 	cmdflags "github.com/weaveworks/flintlock/internal/command/flags"
 	"github.com/weaveworks/flintlock/internal/config"
 	"github.com/weaveworks/flintlock/internal/version"
-	"github.com/weaveworks/flintlock/pkg/flags"
 	"github.com/weaveworks/flintlock/pkg/log"
 )
 
-// NewCommand creates a new cobra command for running the gRPC HTTP gateway.
-func NewCommand(cfg *config.Config) *cobra.Command {
-	cmd := &cobra.Command{
-		Use:   "gw",
-		Short: "Start serving the HTTP gateway for the flintlock gRPC API",
-		PreRunE: func(c *cobra.Command, _ []string) error {
-			flags.BindCommandToViper(c)
-
-			logger := log.GetLogger(c.Context())
-			logger.Infof(
-				"flintlockd, version=%s, built_on=%s, commit=%s",
-				version.Version,
-				version.BuildDate,
-				version.CommitHash,
-			)
-
-			return nil
-		},
-		RunE: func(c *cobra.Command, _ []string) error {
-			return runGWServer(c.Context(), cfg)
+// NewCommand creates a new cli command for running the gRPC HTTP gateway.
+func NewCommand(cfg *config.Config) *cli.Command {
+	cmd := &cli.Command{
+		Name:  "gw",
+		Usage: "Run the gRPC HTTP gateway",
+		Action: func(c *cli.Context) error {
+			return runGWServer(c.Context, cfg)
 		},
 	}
 
 	cmdflags.AddGWServerFlagsToCommand(cmd, cfg)
 
+	cmd.Before = func(context *cli.Context) error {
+		// Load the configuration file
+		configPath := context.String("config")
+		if _, err := os.Stat(configPath); err == nil || context.IsSet("config") {
+			inputSource, err := altsrc.NewYamlSourceFromFile(configPath)
+			if err != nil {
+				return fmt.Errorf("unable to create input source with context: %w", err)
+			}
+
+			err = altsrc.ApplyInputSourceValues(context, inputSource, cmd.Flags)
+			if err != nil {
+				return fmt.Errorf("unable to apply input source with context: %w", err)
+			}
+		}
+
+		logger := log.GetLogger(context.Context)
+		logger.Infof(
```

Comment:
We could probably move the version output to the root command?

Reply:
Actually how does this work? Is there the equivalent of the persistent pre-run?

Reply:
There is no `persistent pre-run` or `persistent flag`. `App.Before` does what `pre-run` does. And we have access to parent flags through the context.

Now I think moving this to root command (app) is a good idea. But it will be shown only when calling that command

Reply:
`App.before` on the App root will run on any subcommand right?

```
	// An action to execute before any subcommands are run, but after the context is ready
	// If a non-nil error is returned, no subcommands are run
	Before BeforeFunc
```
https://github.com/urfave/cli/blob/v1.22.5/app.go#L28

Or is that what you're saying? Maybe I'm not understanding

Reply:
I'm sorry I forgot the context of this comment. But yeah `App.before` always run.  A subCommand can also declare `command.Before`.